apps/client/src/app/core/data/lazy-data-provider.service.ts

```diff
@@ -0,0 +1,46 @@
+import { isPlatformServer } from '@angular/common';
+import { HttpClient } from '@angular/common/http';
+import { Inject, Injectable, PLATFORM_ID } from '@angular/core';
+import { environment } from 'apps/client/src/environments/environment';
+import { Observable } from 'rxjs';
+import { shareReplay, take } from 'rxjs/operators';
+import { PlatformService } from '../tools/platform.service';
+import { LazyData, LazyDataKey } from './lazy-data';
+import { lazyFilesList } from './lazy-files-list';
+
+/**
+ * A service providing cached observables that resolve lazy data.
+ */
+@Injectable({ providedIn: 'root' })
+export class LazyDataProviderService {
+  private readonly lazyLoaderCache = new Map<LazyDataKey, Observable<LazyData[LazyDataKey]>>();
+
+  constructor(private readonly http: HttpClient, private readonly platformService: PlatformService, @Inject(PLATFORM_ID) private readonly platform: Object) {}
+
+  /**
+   * Gets an observable that will resolve lazy data for the specified key.
+   * For each key, only a single `ReplaySubject`-like observable will ever be created.
+   * This ensures that each key will only ever be loaded once, but that loading will be
+   * defered until the observable is actually subscribed to.
+   *
+   * @param key The key of the lazy data to retrieve.
+   * @returns An observable that will emit the lazy data associated with the specified key, then completes.
+   */
+  public readonly getLazyData = <T extends LazyDataKey>(key: T): Observable<LazyData[T]> => {
```

Comment:
It is just me being defensive with functions that might be used as arguments to other functions. Typescript doesn't implicitly defend against unbound `this` (unless the angular bundler is doing some magic on top of it that I'm unaware of).

Reply:
Ok I see, I think it's better to have them as functions right now, because there's pretty much nothing that takes a function as parameter in TC as I'm always going for rxjs over callbacks, and the only functions created on the fly are pure.